core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -442,76 +498,101 @@ sealed trait AsyncRequest {
+case class DeleteRequest(path: String, version: Int, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = DeleteResponse
 }
 
-case class ExistsRequest(path: String, ctx: Option[Any] = None) extends AsyncRequest {
+case class ExistsRequest(path: String, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = ExistsResponse
 }
 
-case class GetDataRequest(path: String, ctx: Option[Any] = None) extends AsyncRequest {
+case class GetDataRequest(path: String, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = GetDataResponse
 }
 
-case class SetDataRequest(path: String, data: Array[Byte], version: Int, ctx: Option[Any] = None) extends AsyncRequest {
+case class SetDataRequest(path: String, data: Array[Byte], version: Int, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = SetDataResponse
 }
 
-case class GetAclRequest(path: String, ctx: Option[Any] = None) extends AsyncRequest {
+case class GetAclRequest(path: String, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = GetAclResponse
 }
 
-case class SetAclRequest(path: String, acl: Seq[ACL], version: Int, ctx: Option[Any] = None) extends AsyncRequest {
+case class SetAclRequest(path: String, acl: Seq[ACL], version: Int, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = SetAclResponse
 }
 
-case class GetChildrenRequest(path: String, ctx: Option[Any] = None) extends AsyncRequest {
+case class GetChildrenRequest(path: String, ctx: Option[Any] = None, checkInfo: Option[ZkCheckInfo] = None) extends AsyncRequest {
   type Response = GetChildrenResponse
 }
 
 sealed abstract class AsyncResponse {
   def resultCode: Code
   def path: String
   def ctx: Option[Any]
+  def checkResult: Option[ZkCheckResult]
 
   /** Return None if the result code is OK and KeeperException otherwise. */
-  def resultException: Option[KeeperException] =
-    if (resultCode == Code.OK) None else Some(KeeperException.create(resultCode, path))
+  def resultException: Option[KeeperException] = {
+    if (resultCode == Code.OK)
+      None
+    else if (checkResult.isEmpty || checkResult.get.resultCode == Code.OK)
+      Some(KeeperException.create(resultCode, path))
+    else
+      Some(checkResult.get.resultException)
+  }
+
 
   /**
    * Throw KeeperException if the result code is not OK.
    */
   def maybeThrow(): Unit = {
     if (resultCode != Code.OK)
-      throw KeeperException.create(resultCode, path)
+      if (checkResult.isEmpty || checkResult.get.resultCode == Code.OK)
+        throw KeeperException.create(resultCode, path)
+      else
+        throw checkResult.get.resultException
   }
 
   def metadata: ResponseMetadata
 }
 
+case class ZkCheckResult(path: String, resultCode: Code) {
```

Comment:
The path is needed to generate keeper exception with path information without assuming it to be controller epoch path. If we are going to generate the `ControllerEpochZkVersionMismatch` exception in `KafkaZkClient` instead of `KeeperException` in `ZookeeperClient`, I think we can remove the path and just keep the resultCode.